rustls/src/conn/unbuffered.rs

```diff
@@ -0,0 +1,442 @@
+                    };
+
+                match self.core.process_msg(msg, state) {
+                    Ok(new) => state = new,
+
+                    Err(e) => {
+                        self.core.state = Err(e.clone());
+                        return Err(e);
+                    }
+                }
+
+                self.core.state = Ok(state);
+            } else if self.wants_write {
+                break (
+                    buffer.pending_discard(),
+                    MustTransmitTlsData { conn: self }.into(),
+                );
+            } else if self
+                .core
+                .common_state
+                .has_received_close_notify
+            {
+                break (buffer.pending_discard(), ConnectionState::ConnectionClosed);
+            } else if self
+                .core
+                .common_state
+                .may_send_application_data
+            {
+                break (
+                    buffer.pending_discard(),
+                    ConnectionState::TrafficTransit(MayEncryptAppData { conn: self }),
+                );
+            } else {
+                break (buffer.pending_discard(), ConnectionState::NeedsMoreTlsData);
+            }
+        };
+
+        Ok(UnbufferedStatus { discard, state })
+    }
+}
+
+/// The current status of the `UnbufferedConnection*`
+#[must_use]
+pub struct UnbufferedStatus<'c, 'i, Data> {
+    /// Number of bytes to discard
+    ///
+    /// After the `state` field of this object has been handled, `discard` bytes must be
+    /// removed from the *front* of the `incoming_tls` buffer that was passed to
+    /// the [`UnbufferedConnectionCommon::process_tls_records`] call that returned this object.
+    ///
+    /// This discard operation MUST happen *before*
+    /// [`UnbufferedConnectionCommon::process_tls_records`] is called again.
+    pub discard: usize,
+
+    /// The current state of the handshake process
+    ///
+    /// This value MUST be handled prior to calling
+    /// [`UnbufferedConnectionCommon::process_tls_records`] again. See the documentation on the
+    /// variants of [`ConnectionState`] for more details.
+    pub state: ConnectionState<'c, 'i, Data>,
+}
+
+/// The state of the [`UnbufferedConnectionCommon`] object
+#[non_exhaustive] // for forwards compatibility; to support caller-side certificate verification
+pub enum ConnectionState<'c, 'i, Data> {
+    /// One, or more, application data records are available
+    ///
+    /// See [`AppDataAvailable`] for more details on how to use the enclosed object to access
+    /// the received data.
+    AppDataAvailable(AppDataAvailable<'c, 'i, Data>),
+
+    /// Connection has been cleanly closed by the peer
+    ConnectionClosed,
+
+    /// A Handshake record is ready for encoding
+    ///
+    /// Call [`MustEncodeTlsData::encode`] on the enclosed object, providing an `outgoing_tls`
+    /// buffer to store the encoding
+    MustEncodeTlsData(MustEncodeTlsData<'c, Data>),
```

Comment:
I don't love these variant (and state type) names. In particular:

* Different style (`AppDataAvailable` is a state, `MustTransmitTlsData` describes the next action)
* Repetition (`ConnectionState::ConnectionClosed`)
* Verbosity (`MustTransmitTlsData`)
* Unclearly linked to connection phases (handshaking vs traffic)

I'd propose something more like this:

* `EncodeHandshake`
* `TransmitHandshake`
* `HandshakeBlocked`
* `ReadTraffic`
* `WriteTraffic`
* `Closed`

(It's also weird to me that `TransmitHandshake` is a separate `ConnectionState` rather than a method on some result of a method on `EncodeHandshake`?)

Reply:
I don't think `EncodeHandshake` and `TransmitHandshake` are specific to handshake messages, eg they are also required for non-handshake messages (both during the handshake, and after).

Reply:
I renamed the variants.